src/exception.c

```diff
@@ -21,13 +27,26 @@
  */
 
 /** @brief Exception handler currently registered with exception system */
-static void (*__exception_handler)(exception_t*) = NULL;
+static void (*__exception_handler)(exception_t*) = exception_default_handler;
 /** @brief Base register offset as defined by the interrupt controller */
 extern const void* __baseRegAddr;
 
 /**
  * @brief Register an exception handler to handle exceptions
  *
+ * The registered handle is responsible for clearing any bits that may cause
+ * a re-trigger of the same exception and updating the EPC. To manipulate the
+ * registers, update the values in the exception_t struct. They will be
+ * restored to appropriate locations when returning from the handler. Setting
+ * them directly will not work as expected as thhey will get overwritten.
+ * An important example is the cause bits (12-17) of FCR31 from cop1. To
+ * prevent re-triggering the exception they should be cleared by the handler.
+ * k0 ($26), k1 ($27), s0-s7 ($16-$23), and gp ($28) are not saved/restored
+ * and will not be available in the handler. These are removed to gain some
+ * performance as they are already saved by GCC when necessary. Theoretically
```

Comment:
I think that for *interrupts* it makes sense not to save those registers as they are callee-saved (BTW, the same I think it applies to f15-f31 but should be verified). Instead for exceptions, one might want to have a look at the whole state of all regs. So maybe the interrupt handlers should save/restore those only after it has verified that it's an exception.

Anyway, this can be improved in the future, no need to do it now.

Reply:
https://github.com/DragonMinded/libdragon/issues/129